DE-98/DE-99 - system_setup: Wsl oobe auto reconf mode

patrick330602 opened the pull request:
This fixes TODO 2 and TODO 3 (with the last commit): 
2. Implement an API in the server to know if we are in reconfigure or setup mode based on existing UID user >= 1000. Fetch that API from the client and set the variant to wsl_setup or wsl_configuration.
3. In wsl_configuration mode the UI shows ”base”/”advanced” screens. (required models list is already updated automatically)

didrocks commented on subiquity/common/apidef.py:
```diff
@@ -77,6 +77,8 @@ def POST(endpoint_names: List[str]) -> None:
             def POST(variant: str) -> None:
                 """Choose the install variant - desktop/server"""
+            def GET() -> str:
+                """Get the install variant - desktop/server"""
```
Please add wsl_setup and wsl_reconfigure.

patrick330602 replied:
This is fixed.

didrocks commented on subiquity/common/apidef.py:
```diff
@@ -77,6 +77,8 @@ def POST(endpoint_names: List[str]) -> None:
         class client_variant:
             def POST(variant: str) -> None:
                 """Choose the install variant - desktop/server"""
```
Please add wsl_setup and wsl_reconfigure in the comment.

didrocks commented on subiquity/server/server.py:
```diff
@@ -117,6 +117,9 @@ async def client_variant_POST(self, variant: str) -> None:
         self.app.base_model.set_source_variant(variant)
 
+    async def client_variant_GET(self) -> str:
+        return self.app.base_model.client_variant
```
I would not store it on the base_model (above), but rather on the application directly. Then, the method is used to adapt the base_model, without storing it.

patrick330602 replied:
Updated.

didrocks commented on subiquity/models/subiquity.py:
```diff
@@ -105,6 +105,8 @@ def __init__(self, root, install_model_names, postinstall_model_names):
         if root != '/':
             self.target = root
 
+        self.client_variant = ''
```
see comment on server.py. I don’t think we should store the variant here but on the application from client_variant_POST.

didrocks commented on system_setup/client/client.py:
```diff
@@ -38,16 +39,10 @@ class SystemSetupClient(SubiquityClient):
+        if is_reconfigure(opts.dry_run):
+            self.variant = "wsl_configuration"
             self.controllers = [
+                "Welcome",
```
We shouldn’t have the Welcome controller loaded in reconfigure mode (see the spec)

patrick330602 replied:
Removed.

didrocks commented on system_setup/client/client.py:
```diff
@@ -38,16 +39,10 @@ class SystemSetupClient(SubiquityClient):
-        # 5. Add language page
-        # self.variant = "wsl_configuration"
-        if opts.reconfigure:
+        if is_reconfigure(opts.dry_run):
```
We shouldn’t detect reconfiguration mode in the client. The model is basically:

1. Let the server detects based on UID rules if we are in setup or reconfiguration mode.
2. Get it from the server API in the client in which mode we are
3. A little bit silly with the current implementation, but needed… Post the mode back to the server for it to load the expected server models…

Even if this seems counter-intuitive, because of the 3. this allows:
- the server to do the initial detection, as this is where the actual code is running (client could be remote)
- let the client override it if needed (for the tests for instance)

patrick330602 replied:
I am not sure how should I do this since we should then call an async function in __init__, which of course crashes. What is the proper way I should implement this?

didrocks replied:
This is a good question as this is triggered by the fact of having static class member with the base inherited/+meta class making magic on controllers, replacing the objects. I think of multiple hackish solutions, but maybe @mwhudson or @dbungert has a more upstream-compatible idea?

mwhudson replied:
So I guess there are two broad approaches here.

1. Have all controllers loaded in the client always but have the ones that are inappropriate for the current mode skip themselves (the generic tui client code has support for this, just raise subiquitycore.tuicontroller.Skip from make_ui)
2. Change the set of controllers to be loaded after seeing what the server says

The first is probably easier, tbh, although the second might be aesthetically more pleasing. Nothing in the installer tui client touches the controllers instance until the first status API call has returned so while the ControllerSet class wasn't written with this in mind it looks like assigning `controller_names` between the instance being constructed and calling `load_all` will work fine (another approach would be to construct the ControllerSet later, probably).

mwhudson replied:
WRT the POSTing the variant back... surely something in the system_state server startup code can call set_source_variant on the model during startup?

patrick330602 replied:
Updated, and tested locally. It should now work properly.

didrocks commented on system_setup/client/controllers/summary.py:
```diff
@@ -87,10 +89,21 @@ async def _wait_status(self, context):
+        if tmp_variant == "wsl_configuration":
+            tmp_uid = os.getuid()
+            if tmp_uid >= 1000 and tmp_uid < 65534:
+                real_name = pwd.getpwuid(tmp_uid)[0]
```
I think we shuld rather change the summary message without using the user name here. Then the whole variant os.getuid is not needed here as we don’t need real_name.

patrick330602 replied:
The logic is changed for this, needs check whether should be improved.

didrocks replied:
I don’t see any changes there, please have a look before I make a second review pass.

patrick330602 replied:
Sorry for that, fixed now

didrocks commented on system_setup/client/controllers/wslconfbase.py:
```diff
@@ -12,7 +12,11 @@ class WSLConfigurationBaseController(SubiquityTuiController):
+        onsite_variant = getattr(self.app, "variant")
+        if onsite_variant is not None:
+            variant = onsite_variant
+        return WSLConfigurationBaseView(self, data, variant)
```
Same, variant should not be needed in the Base View, it’s the same in both setup and reconfiguration

patrick330602 replied:
This one is for In wsl_configuration mode the UI shows ”base”/”advanced” screens.

didrocks replied:
See my other comment below, which should make this uneeded :)

patrick330602 replied:
addressed

didrocks commented on system_setup/common/helpers.py:
```diff
@@ -0,0 +1,31 @@
+            # check every normal user except nobody (65534)
+            if int(line.split(':')[2]) >= 1000 and \
+               int(line.split(':')[2]) != 65534:
+                count += 1
```
Rather than counting the user, you can stop as soon as you have found one matching user and return true

didrocks commented on system_setup/common/helpers.py:
```diff
@@ -0,0 +1,31 @@
+               int(line.split(':')[2]) != 65534:
+                count += 1
+    is_none_dryrun_normaluser = not is_dryrun and count != 0
+    return is_dryrun_reconfigure or is_none_dryrun_normaluser
```
Return early in the function if is_dryrun_reconfigure is true just below DRYRUN_RECONFIG environment get.

didrocks commented on system_setup/ui/views/wslconfbase.py:
```diff
@@ -95,9 +95,11 @@ class WSLConfigurationBaseView(BaseView):
-
+        self.is_conf = variant == "wsl_configuration"
+        if self.is_conf:
+            self.title = _("WSL configuration - Base options")
```
We should always use that title without using the variant.

patrick330602 replied:
This one is for In wsl_configuration mode the UI shows ”base”/”advanced” screens though, People would be confused that there are two configuration pages with the same title, otherwise we should change the description

didrocks replied:
Yes. but my suggestion was to always use that title rather. Those are the base options and can show them as such in setup mode.

patrick330602 replied:
Addressed

didrocks commented:
Something else I didn’t spot yesterday: please target `main` branch instead of `wsl_oobe_reduce_dup`. This has no dependency on `wsl_oobe_reduce_dup` and should not include this commit. As you changed `wsl_oobe_reduce_dup`, you would need otherwise to due a useless rebase (and same later for other branches that depends on this one)

didrocks commented:
There is still one non outdated comment that needs to be fixed IMHO and some merge conflicts, do you mind fixing those before requesting a new review?

didrocks reviewed (changes requested):
Some changes that were requested in this PR are still not here. Also, the branch needs a rebase and not a merge as already stated.

patrick330602 commented:
@didrocks I just checked and I have fixed everything that was requested, including rebase to the main. Am I missing anything here?

didrocks commented:
From what I see in a quick check:
- you merged main, not do a `git rebase`
- WSLConfigurationAdvanced still inherites from WSLConfigurationBase (we removed it and asked to remove it multiple times, is the diff not up to date with your local checking? :/)
- There are some comments added like
`# controllers = {
    #    "server": [
`
- In the diff, there are some empty lines added/removed

Please do a rebase, and then do a real serious check that all previously changes are fixed, that there is no leftover of debug and no uneeded diffs.

patrick330602 commented:
> From what I see in a quick check:
> 
> * you merged main, not do a `git rebase`
> * WSLConfigurationAdvanced still inherites from WSLConfigurationBase (we removed it and asked to remove it multiple times, is the diff not up to date with your local checking? :/)
> * There are some comments added like
> 
> ```
> # controllers = {
>     #    "server": [
> ```
> 
> * In the diff, there are some empty lines added/removed
> 
> Please do a rebase, and then do a real serious check that all previously changes are fixed, that there is no leftover of debug and no uneeded diffs.

Updated and fixed. Could you check it again? Thanks!

didrocks commented on subiquity/client/client.py:
```diff
@@ -112,7 +112,9 @@ def make_ui(self):
         "SSH",
         "SnapList",
         "Progress",
-        ]
```
Remove this ] in diff and keep existing alignement.

didrocks commented on subiquity/server/server.py:
```diff
@@ -247,6 +251,7 @@ def make_model(self):
 
     def __init__(self, opts, block_log_dir):
         super().__init__(opts)
+        self.variant = self.supported_variants[0]
```
This should call `self.set_source_variant()`

didrocks commented on subiquitycore/core.py:
```diff
@@ -32,13 +32,15 @@ class Application:
     # project = "subiquity"
-    # controllers = [
+    # controllers = {
+    #    "server": [
```
Please remove all this file diff, not correct anymore with the current implementation (this was already a comment in the previous review)

didrocks commented on system_setup/ui/views/summary.py:
```diff
@@ -12,7 +12,7 @@
 from subiquitycore.ui.buttons import (
     cancel_btn,
     ok_btn,
-    )
```
Remove this and keep existing alignement.

didrocks commented on system_setup/ui/views/summary.py:
```diff
@@ -60,7 +64,7 @@ def update_for_state(self, state):
             btns = [
                 self.view_error_btn,
                 self.reboot_btn,
-                ]
```
Remove this and keep existing alignement.

didrocks commented on system_setup/ui/views/wslconfbase.py:
```diff
@@ -97,7 +97,6 @@ class WSLConfigurationBaseView(BaseView):
 
     def __init__(self, controller, configuration_data):
         self.controller = controller
-
```
Remove this not relevant in the current PR.

didrocks reviewed (changes requested):
Some small typos and alignements fixes needed. Then, it will be ready for upstream to review.

didrocks commented:
You should remove https://github.com/canonical/subiquity/pull/1042/commits/e0884e9943c23c21994c8cc4bedd50f8e87fa612. This commit doesn’t do anything apart from introducing a diff.

didrocks commented:
You reintroduce the inheritance in https://github.com/canonical/subiquity/pull/1042/commits/675e6afeec3d9e59e194492810cd9752b28cf36c. Please fix this and remove last commit.

didrocks commented:
Ok, all comments are now resolved, the tests pass. @mwhudson @dbungert, mind having a review of the non system_setup part?

mwhudson commented on system_setup/server/controllers/configure.py:
```diff
@@ -33,6 +33,7 @@ class ConfigureController(SubiquityController):
 
     def __init__(self, app):
         super().__init__(app)
+        self.app = app
```
This line isn't necessary, surely?

mwhudson replied:
Well OK we can argue about how and whether it's needed in that branch then :)

mwhudson commented on system_setup/client/client.py:
```diff
@@ -36,21 +37,43 @@ class SystemSetupClient(SubiquityClient):
         ]
+    }
+
+    def restart(self, remove_last_screen=True, restart_server=False):
```
Wait, didn't I ask you to parameterize the superclass implementation of this method in a previous PR?

patrick330602 replied:
I am not clear about this.... could you elaborate more?

mwhudson replied:
Having read things a bit more, I don't think you need to add the restart method to this class at all. Can you explain why you did that?

patrick330602 replied:
Now I check again, yep you are correct, the restart is not needed here.

patrick330602 replied:
done for the fixes here.

mwhudson commented on subiquity/server/server.py:
```diff
@@ -123,6 +123,10 @@ async def client_variant_POST(self, variant: str) -> None:
             self.app.controllers.Source.configured()
         else:
             self.app.base_model.set_source_variant(variant)
+        self.app.variant = variant
```
It would be more consistent for this to call self.app.set_source_variant would it not?

patrick330602 replied:
done for the fixes

mwhudson reviewed (commented):
Seems ok, a few questions. Happy for Dan to merge this if you can convince him your responses to my questions are ok :)

mwhudson reviewed (approved):
LGTM, thanks for the changes
